### ISSUE TITLE
windows test.yml fails on Chinese windows

### ISSUE THREAD

**Issue by rophy** [ansible/ansible#11198]
##### Issue Type:

Bug Report
##### Ansible Version:

[root@lsc102 ansible-examples]# ansible --version
ansible 2.0.0 (devel 8cf4452d48) last updated 2015/05/04 15:01:24 (GMT +800)
  lib/ansible/modules/core: (detached HEAD 9028e9d4be) last updated 2015/05/04 15:01:44 (GMT +800)
  lib/ansible/modules/extras: (detached HEAD dd80fa221c) last updated 2015/05/04 15:02:00 (GMT +800)
  v2/ansible/modules/core: (detached HEAD 0341ddd35e) last updated 2015/05/04 15:02:16 (GMT +800)
  v2/ansible/modules/extras: (detached HEAD dd80fa221c) last updated 2015/05/04 15:02:37 (GMT +800)
  configured module search path = None
##### Ansible Configuration:

Clean "yum install ansible" on Fedora 19
##### Environment:

Ansible Host: fedora 19
Ansible Target: Windows 2008 R2 SP1 - Chinese (Hong Kong)
##### Summary:

Running ansible-examples/windows/test.yml fails, I believe it's due to "ipconfig" displaying Chinese characters. (target windows has default chcp: 950)

UnicodeDecodeError: 'ascii' codec can't decode byte 0xb0 in position 49: ordinal not in range(128)
##### Steps To Reproduce:
1. Change windows system locale to "Chinese (Hong Kong)" (requires reboot)
2. In a command line, run "ipconfig" and see that its output should be in Chinese (chcp 950)
3. [root@lsc102 ansible-examples]# ansible-playbook  windows/test.yml
##### Expected Results:

I expect the execution to pass.
##### Actual Results:

``` shell

[root@lsc102 ansible-examples]# ansible-playbook  windows/test.yml

PLAY [test raw module] ********************************************************

GATHERING FACTS ***************************************************************
ok: [lsc624.tw.ibm.com]

TASK: [run ipconfig] **********************************************************
fatal: [lsc624.tw.ibm.com] => Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 590, in _executor
    exec_rc = self._executor_internal(host, new_stdin)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 792, in _executor_internal
    return self._executor_internal_inner(host, self.module_name, self.module_args, inject, port, complex_args=complex_args)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 1025, in _executor_internal_inner
    result = handler.run(conn, tmp, module_name, module_args, inject, complex_args)
  File "/usr/lib/python2.7/site-packages/ansible/runner/action_plugins/raw.py", line 47, in run
    su=self.runner.su)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 1173, in _low_level_exec_command
    in_data=in_data)
  File "/usr/lib/python2.7/site-packages/ansible/runner/connection_plugins/winrm.py", line 153, in exec_command
    return (result.status_code, '', result.std_out.encode('utf-8'), result.std_err.encode('utf-8'))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xb0 in position 49: ordinal not in range(128)


FATAL: all hosts have already failed -- aborting

PLAY RECAP ********************************************************************
           to retry, use: --limit @/root/test.retry

lsc624.tw.ibm.com          : ok=1    changed=0    unreachable=1    failed=0

```

Note: the ipconfig task will success if I change the system locale back to English (United States)

**Comment by abadger** [comment]
Since we're changing that output into a byte str instead of a unicode type string there could be some deeper issues lurking here but the fix for that specific traceback should be pretty simple.  Let me work up a fix.

**Comment by abadger** [comment]
@rophy I don't have a windows box to test on; if you please test that commit to see that it solves your issue I would appreciate it.  Thanks!

I'll close this in a few days if there's no report that this is still broken.

**Comment by rophy** [comment]
No problem, let me test the fix and get back to you (I just need to pull latest changes to test it, right?)

**Comment by abadger** [comment]
Correct.

**Comment by cchurch** [comment]
@abadger I'd attempted a similar fix via https://github.com/ansible/ansible/pull/10024 (needs rebase), should be able to close that one if this fix works.  I'll test also.

**Comment by rophy** [comment]
I couldn't get winrm to work after updating to latest trunk, which blocks me from verifying this fix... I have opened another issue #11217.

**Comment by rophy** [comment]
ok, winrm works if I run "source v1/hacking/env-setup" instead of "source hacking/env-setup".

however,  it seems the fix didn't solve the problem.

```
[root@lsc102 ansible]# ansible windows -m raw -a ipconfig -vvvv
<lsc624.tw.ibm.com> ESTABLISH WINRM CONNECTION FOR USER: Administrator on PORT 5986 TO lsc624.tw.ibm.com
<lsc624.tw.ibm.com> WINRM CONNECT: transport=plaintext endpoint=https://lsc624.tw.ibm.com:5986/wsman
<lsc624.tw.ibm.com> EXEC ipconfig
<lsc624.tw.ibm.com> WINRM EXEC 'ipconfig' []
<lsc624.tw.ibm.com> WINRM RESULT <Response code 0, out " Windows IP Configu", err "">
lsc624.tw.ibm.com | FAILED => Traceback (most recent call last):
  File "/root/ansible/v1/ansible/runner/__init__.py", line 582, in _executor
    exec_rc = self._executor_internal(host, new_stdin)
  File "/root/ansible/v1/ansible/runner/__init__.py", line 785, in _executor_internal
    return self._executor_internal_inner(host, self.module_name, self.module_args, inject, port, complex_args=complex_args)
  File "/root/ansible/v1/ansible/runner/__init__.py", line 1032, in _executor_internal_inner
    result = handler.run(conn, tmp, module_name, module_args, inject, complex_args)
  File "/root/ansible/v1/ansible/runner/action_plugins/raw.py", line 47, in run
    become=self.runner.become)
  File "/root/ansible/v1/ansible/runner/__init__.py", line 1169, in _low_level_exec_command
    in_data=in_data)
  File "/root/ansible/v1/ansible/runner/connection_plugins/winrm.py", line 165, in exec_command
    return (result.status_code, '', result.std_out.encode('utf-8'), result.std_err.encode('utf-8'))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xb0 in position 49: ordinal not in range(128)

[root@lsc102 ansible]# ansible --version
ansible 2.0.0 (devel 757fb39a2e) last updated 2015/06/10 12:15:29 (GMT +800)
  lib/ansible/modules/core: (detached HEAD d6ed6113a7) last updated 2015/06/10 12:15:50 (GMT +800)
  lib/ansible/modules/extras: (detached HEAD 57813a2e74) last updated 2015/06/10 12:16:11 (GMT +800)
  v1/ansible/modules/core: (detached HEAD f8d8af17cd) last updated 2015/06/10 12:16:30 (GMT +800)
  v1/ansible/modules/extras: (detached HEAD 495ad450e5) last updated 2015/06/10 12:16:51 (GMT +800)
  configured module search path = None


```

**Comment by rophy** [comment]
In case it might matters:  my experiences in python + non-English windows tells me that windows codepage (chcp) plays a big part in command line outputs, and in many cases, it's not as simple as converting between str and unicode type.

for example, in Chinese (Hong Kong), "ipconfig" in command line will be encoded in chcp 950 (Big-5 codepage), and if python isn't aware that the codepage is 950, attempting to convert it to unicode will result in UnicodeDecodeError.

**Comment by rophy** [comment]
I did some testing for the #10024 fix.
When pywinrm protocol uses the default codepage (cp437), it essentially ignores target machine's system locale, and make the output of ipconfig into english. So the fix in #10024 would solve the error, but the ipconfig outputs will be in English (where in native system locale it should be in Chinese).

Sample outputs below. Notice that the Ethernet adapter has a Chinese name, and with cp437 it becomes corrupted:

```
[root@lsc102 ansible]# ansible windows -m raw -a ipconfig -vvvv
<lsc624.tw.ibm.com> ESTABLISH WINRM CONNECTION FOR USER: Administrator on PORT 5986 TO lsc624.tw.ibm.com
<lsc624.tw.ibm.com> WINRM CONNECT: transport=plaintext endpoint=https://lsc624.tw.ibm.com:5986/wsman
<lsc624.tw.ibm.com> EXEC ipconfig
<lsc624.tw.ibm.com> WINRM EXEC 'ipconfig' []
<lsc624.tw.ibm.com> WINRM RESULT <Response code 0, out " Windows IP Configu", err "">
lsc624.tw.ibm.com | success | rc=0 >>

Windows IP Configuration


Ethernet adapter ░╧░∞│s╜u:

   Connection-specific DNS Suffix  . : tw.ibm.com
   Link-local IPv6 Address . . . . . : fe80::25db:305a:e878:20a5%11
   IPv4 Address. . . . . . . . . . . : 172.16.148.118
   Subnet Mask . . . . . . . . . . . : 255.255.248.0
   Default Gateway . . . . . . . . . : 172.16.151.254

Tunnel adapter isatap.tw.ibm.com:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . : tw.ibm.com



```

If I manually specify codepage=950 in "v1/ansible/runner/connection_plugins/winrm.py" line 122:

```
        if not self.shell_id:
            self.shell_id = self.protocol.open_shell(codepage=950)
        command_id = None

```

and then in "v1/ansible/runner/connection_plugins/winrm.py" line 165:

```
        return (result.status_code, '', result.std_out.decode('cp950'), result.std_err.decode('cp950'))
```

then I would get Chinese ipconfig output correctly (along with the Chinese adpater name):

```
[root@lsc102 ansible]# ansible windows -m raw -a ipconfig -vvvv
<lsc624.tw.ibm.com> ESTABLISH WINRM CONNECTION FOR USER: Administrator on PORT 5986 TO lsc624.tw.ibm.com
<lsc624.tw.ibm.com> WINRM CONNECT: transport=plaintext endpoint=https://lsc624.tw.ibm.com:5986/wsman
<lsc624.tw.ibm.com> EXEC ipconfig
<lsc624.tw.ibm.com> WINRM EXEC 'ipconfig' []
<lsc624.tw.ibm.com> WINRM RESULT <Response code 0, out " Windows IP ▒]▒w ", err "">
lsc624.tw.ibm.com | success | rc=0 >>

Windows IP 設定


乙太網路卡 區域連線:

   連線特定 DNS 尾碼 . . . . . . . . : tw.ibm.com
   連結-本機 IPv6 位址 . . . . . . . : fe80::25db:305a:e878:20a5%11
   IPv4 位址 . . . . . . . . . . . . : 172.16.148.118
   子網路遮罩 . . . . . . . . . . . .: 255.255.248.0
   預設閘道 . . . . . . . . . . . . .: 172.16.151.254

通道介面卡 isatap.tw.ibm.com:

   媒體狀態 . . . . . . . . . . . . .: 媒體已中斷連線
   連線特定 DNS 尾碼 . . . . . . . . : tw.ibm.com

```

**Comment by abadger** [comment]
Yeah, I didn't apply any changes to the v1 tree, only to the devel tree.  If you ""source v1/hacking/env-setup" then you'll get the old code instead of the code that I fixed.

My fix shouldn't traceback (in the same place, at least... could traceback further down the line) as long as python itself isn't attempting to decode the output of ipconfig (I've never tried to understand the windows specific portions of subprocess but I don't think it would do that).  However, it may yield mangled output.  I think that with the present code it won't as to_bytes() should be a no-op here.... subprocess should be returning a byte str and to_bytes() will pass that through unmodified.  We'll have to get winrm working in the current devel branch to test this fix.

Looking at #10024 I hate to hardcode the codepage but since pywinrm seems to default to 437 it's probably correct to use that here too (rather than utf-8, to_bytes()'s default).  Maybe we should be explicit about the codepage when we call open_shell() as well.  that way we will be better prepared to change both places if we need to do something other than cp437.  Also, the other connection plugins are currently returning byte str.... We should either convert all of them to return unicode strings or make sure that this one is a byte str when we return it.  So maybe something like this:

``` python
[...]
            # Must match with the encoding that we use to normalize strings when converting to unicode type strings
            self.shell_id = self.protocol.open_shell(codepage=437)
[...]
        # Must match with the codepage used when we call open_shell()
        std_out = to_unicode(result.std_out.decode, 'cp437')
        std_err = to_unicode(result.std_err.decode, 'cp437')
        std_out = to_bytes(std_out, 'utf-8')
        std_err = to_bytes(std_out, 'utf-8')
        return (result.status_code, '', std_out, std_err)
```
